### ISSUE TITLE
Inline Whiteboard

### ISSUE THREAD

**Issue by pylang** [jupyterlab/jupyterlab#9194]
### Problem

I'm pleased with Jupyter notebooks and the new momentum to make JupyterLab more notebook-like.

Based a [prior recommendation](https://github.com/jupyterlab/jupyterlab/issues/7468#issuecomment-674259279) and reduced activity on the notebook, I link this [issue](https://github.com/jupyter/notebook/issues/1999#issuecomment-713443528).  It is a feature several people seem excited about.  It would be nice to add it here to JupyterLab.

### Proposed Solution

In short, add a whiteboard to the notebook.

![4c12aeb2-c6e1-11e6-9560-61d176a9c2d4](https://user-images.githubusercontent.com/10778668/96718245-bb55dd00-1375-11eb-9d9e-4faf248c93e2.png)

### Additional context

See links above.

**Comment by jasongrout** [comment]
That is an excellent custom extension idea!

**Comment by rukie** [comment]
Looks like ipycanvas could achieve this.
Example using a mouse
https://notebooks.gesis.org/binder/jupyter/user/martinrenou-ipycanvas-451myeai/lab/tree/examples/hand_drawing.ipynb

Developer's twitter video 
https://twitter.com/martinRenou/status/1354121784539414529

And since his example used a mouse, not a pen:
https://stackoverflow.com/questions/54003194/how-can-i-access-tablet-pen-data-via-python

**Comment by telamonian** [comment]
This also seems like an excellent target for the upcoming RTC work (cough, @SylvainCorlay, cough)

**Comment by 1kastner** [comment]
> Example using a mouse
https://notebooks.gesis.org/binder/jupyter/user/martinrenou-ipycanvas-451myeai/lab/tree/examples/hand_drawing.ipynb

For me that link did not work. I guess everybody needs to start their own binder instance? 

I went to https://github.com/martinRenou/ipycanvas/blob/master/examples/hand_drawing.ipynb and yes, the example is awesome!

**Comment by 1kastner** [comment]
@pylang would that ipycanvas hand drawing example suffice your expectations? By having it separately packed, I guess the modularity is even nicer (independent updates etc.) than if it was shipped altogether. Or did I miss one of your requirements?

**Comment by pylang** [comment]
@1kastner  I appreciate your response. :)

I actually don't remember the details of this request.  It's an old request from notebook days.  I think I just wanted to scribble in the notebook and retain them.  I haven't played with ipycanvas enough to see if that will suffice, though I suspect it might.  Having a binder example would be helpful to try it out indeed.

**Comment by 1kastner** [comment]
Well, if you go to https://github.com/martinRenou/ipycanvas you find the launch button.

**Comment by pylang** [comment]
It's a start.  I found it to lag a bit.  I'd be more inclined to use Paint and attach the image.

**Comment by fcollonval** [comment]
Hey for information - the recent tool [Excalidraw](https://github.com/excalidraw/excalidraw) is awesome. The user experience is so much better than any thing I have seen for whiteboard; especially because it is based on the simple principle that we almost always have a keyboard and a mouse - so what's the point to reproduce the way pen & paper works (see author video https://www.youtube.com/watch?v=fix2-SynPGE around 12:45).

**Comment by 1kastner** [comment]
> 
> Hey for information - the recent tool [Excalidraw](https://github.com/excalidraw/excalidraw) is awesome. The user experience is so much better than any thing I have seen for whiteboard; especially because it is based on the simple principle that we almost always have a keyboard and a mouse - so what's the point to reproduce the way pen & paper works (see author video https://www.youtube.com/watch?v=fix2-SynPGE around 12:45).

Can you tell how that solution can be integrated in a Jupyter Notebook? Cool it's under MIT license and can be self-hosted but I guess there is still a way to go create a Jupyter widget that could properly nest it.

**Comment by fcollonval** [comment]
Hey @1kastner indeed, excalidraw will require some development. As they have an export format, the best approach would to use it to create a mime type renderer and a Jupyter Widget; like ipydrawio

**Comment by jasongrout** [comment]
This issue is in the [4.0 planning project](https://github.com/jupyterlab/jupyterlab/projects/12). In the JupyterLab dev meeting this morning, we did not find anyone who was willing to do this by the end of the year to make the 4.0 release. If there is someone willing to work on this, please comment here (and please come to the JLab dev meeting and let us know). If no one has volunteered by next week (10 Nov), we'll take it out of the [4.0 planning project](https://github.com/jupyterlab/jupyterlab/projects/12). If someone volunteers to do it by the end of the year, we can always put it back on the 4.0 roadmap.

**Comment by adpatter** [comment]
@jasongrout We want to implement this.  I'm confident that I can implement this, given some guidance on how we want the finished product to look.  However, I am still pretty ignorant about the new RTC stuff.  Is there anyone who could help me get up to speed on that?

**Comment by fcollonval** [comment]
@adpatter which direction would you like to use? If you are willing to use [Excalidraw](https://www.npmjs.com/package/@excalidraw/excalidraw) - I can support you a bit.

**Comment by adpatter** [comment]
@fcollonval I think that would be a reasonable approach.  One thought we had on this is that the drawing would need to scale to arbitrary screen sizes.  Would a canvas based approach work, given that requirement?

**Comment by fcollonval** [comment]
Testing with Excalidraw app, you can move the drawing area like it is limitless. So I assume it would handle it properly.

**Comment by srkunze** [comment]
not sure what the main discussion thread is - so reposting here:

+1

recently saw this (react component) https://twitter.com/tldraw/status/1461011977791234051

or this

https://github.com/excalidraw/excalidraw

**Comment by srkunze** [comment]
One main issue I have with all the “non-jupyter” solutions is the fact that I would love to see collaboration when using a whiteboard AND thus shared persistence.

Especially this last aspect usually created headaches when it comes to maintenance. So a properly configured JupyterLab/Hub would solve that one for a lot of people.

**Comment by zsunberg** [comment]
This seems to provide the desired functionality: https://github.com/nicknytko/notebook-drawing
